### PR TITLE
Allow insights powershell to unselect data sinks in diagnosticsettings.

### PR DESCRIPTION
**Pull request by vcanaa**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 1]
```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
...
+        private void SetSelectedTimegrains(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
             {
-                if (this.Categories != null)
-                {
-                    foreach (string category in this.Categories)
-                    {
-                        LogSettings logSettings = properties.Logs.FirstOrDefault(x => string.Equals(x.Category, category, StringComparison.OrdinalIgnoreCase));
+                throw new ArgumentException("Parameter 'Enabled' is required by 'Timegrains' parameter.");
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa why not create a parameter set where `Enabled` is required if `Timegrains` is used?

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 2]
```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
...
+                metricSettings.Enabled = this.Enabled.Value;
+            }
+        }
 
-                if (this.Timegrains != null)
+        private void SetSelectedCategories(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
+            {
+                throw new ArgumentException("Parameter 'Enabled' is required by 'Categories' parameter.");
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa why not create a parameter set where `Enabled` is required if `Categories` is used?

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 3]
```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
...
+        {
+            if (!string.IsNullOrWhiteSpace(this.WorkspaceId))
+            {
+                if (this.DisableWorkspace)
                 {
-                    foreach (LogSettings logSettings in properties.Logs)
-                    {
-                        logSettings.RetentionPolicy = retentionPolicy;
-                    }
+                    throw new ArgumentException("WorkspaceId and DisableWorkspace cannot be both present.");
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa use parameter sets to ensure that these parameters are mutually exclusive

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 4]
```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
...
+        {
+            if (!string.IsNullOrWhiteSpace(this.ServiceBusRuleId))
+            {
+                if (this.DisableServiceBus)
                 {
-                    foreach (MetricSettings metricSettings in properties.Metrics)
-                    {
-                        metricSettings.RetentionPolicy = retentionPolicy;
-                    }
+                    throw new ArgumentException("ServiceBusId and DisableServiceBus cannot be both present.");
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa use parameter sets to ensure that these parameters are mutually exclusive

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 5]
```diff
@@ -41,6 +41,27 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
...
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Disable service bus")]
+        [ValidateNotNullOrEmpty]
+        public SwitchParameter DisableServiceBus { get; set; }
+
+        /// <summary>
+        /// Gets or sets the switch for workspace parameter of the cmdlet
+        /// </summary>
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Disable workspace")]
+        [ValidateNotNullOrEmpty]
+        public SwitchParameter DisableWorkspace { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa how do you handle the case of the three Disable parameters being used with the `Enabled` parameter? If a user shouldn't be able to use all four together (or `Enabled` with any combination of the three Disable parameters), then you should consider creating two parameters - one that has `Enabled`, and another that has the three Disable parameters

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 6]
```diff
@@ -41,6 +41,27 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
...
+        /// Gets or sets the switch for eventhub parameter of the cmdlet
+        /// </summary>
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Disable service bus")]
+        [ValidateNotNullOrEmpty]
+        public SwitchParameter DisableServiceBus { get; set; }
+
+        /// <summary>
+        /// Gets or sets the switch for workspace parameter of the cmdlet
+        /// </summary>
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Disable workspace")]
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa please remove the `ValueFromPipelineByPropertyName` attribute from the three `SwitchParameter` parameters you have added

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 7]
```diff
@@ -57,9 +78,9 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
         /// <summary>
         /// Gets or sets the enable parameter of the cmdlet
         /// </summary>
-        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The value indicating whether the diagnostics should be enabled or disabled")]
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The value indicating whether the diagnostics should be enabled or disabled")]
         [ValidateNotNullOrEmpty]
-        public bool Enabled { get; set; }
+        public bool? Enabled { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: @vcanaa when can this be nullable?

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs (hunk 5 above)
Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: The powershell is supposed to offer a way to easily build the request. If the request ends up not being valid, we want to the server to validate and respond with proper message.

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs (hunk 7 above)
Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: Customer can run the following
Set-AzureRmDiagnosticSettings -ResourceId <> -DisableStorage

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs (hunk 1 above)
Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: I studied the parameter set and I didn't find a configuration that satisfies all the requirements. I decided to not use them.

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs (hunk 2 above)
Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: Parameter set doesn't cover all requirements.

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs (hunk 3 above)
Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: Parameter set doesn't cover all requirements.

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs (hunk 4 above)
Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: Parameter set doesn't cover all requirements.

**Comment by shahabhijeet** [comment]
@vcanaa please reopen this PR once you address the feedback and fix build failures.

**Comment by markcowl** [comment]
@azuresdkci retest this please

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1409/

File: src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs [hunk 8]
```diff
@@ -31,6 +32,11 @@ namespace Microsoft.Azure.Commands.Insights.Diagnostics
     [Cmdlet(VerbsCommon.Set, "AzureRmDiagnosticSetting"), OutputType(typeof(PSServiceDiagnosticSettings))]
```
Comment (markcowl) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: Change cmdlet name to update and place an alias to Set (or vice-versa)  Also, please note in upcoming breaking changes that the cmdlet name is changing to update: https://github.com/Azure/azure-powershell/blob/dev/src/ResourceManager/Insights/documentation/upcoming-breaking-changes.md

Comment (vcanaa) [inline src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs]: Can we just leave the name as Set-DiagnsoticSettings? We are improving the api to support multiple diagnostic settings. After that we plan to retire this commandlet and have only Add-AzureRmDiagnosticSettings. I don't think we should change the name a command that is soon to be deprecated.

**Pull request merged**